src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,235 @@
+            ImmutableArray<BoundExpression> rightElements = GetTupleParts(right, temps, effects, ref ignored);
+
+            Debug.Assert(operators.NestedOperators.Length == leftElements.Length && leftElements.Length == rightElements.Length);
+
+            BoundExpression currentResult = null;
+            for (int i = 0; i < leftElements.Length; i++)
+            {
+                var nextLogicalOperand = RewriteTupleOperator(operators.NestedOperators[i], leftElements[i], rightElements[i], type, temps, effects, logicalOperator);
+                if (currentResult is null)
+                {
+                    currentResult = nextLogicalOperand;
+                }
+                else
+                {
+                    currentResult = _factory.Binary(logicalOperator, type, currentResult, nextLogicalOperand);
+                }
+            }
+
+            return currentResult;
+        }
+
+        // TODO factor this method
+        private ImmutableArray<BoundExpression> GetTupleParts(BoundExpression tuple,
+            ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> effects, ref bool wasNotLiteral)
+        {
+            // Example:
+            // (1, 2) == (1, 2);
+            if (IsTupleExpression(tuple.Kind))
+            {
+                return ((BoundTupleExpression)tuple).Arguments;
+            }
+
+            // Example:
+            // (1L, 2L == (1, 2);
+            // (1, "hello") == (1, null);
+            if (tuple.Kind == BoundKind.Conversion)
+            {
+                var tupleConversion = (BoundConversion)tuple;
+                if (tupleConversion.Conversion.Kind == ConversionKind.ImplicitNullable)
+                {
+                    return GetTupleParts(tupleConversion.Operand, temps, effects, ref wasNotLiteral);
+                }
+
+                if ((tupleConversion.Conversion.Kind == ConversionKind.ImplicitTupleLiteral || tupleConversion.Conversion.Kind == ConversionKind.Identity)
+                    && IsTupleExpression(tupleConversion.Operand.Kind))
+                {
+                    return ((BoundTupleExpression)tupleConversion.Operand).Arguments;
+                }
+            }
+
+            // Example:
+            // t == GetTuple();
+            // t == ((byte, byte)) (1, 2);
+            // t == ((short, short))((int, int))(1L, 2L);
+            if (tuple.Type.IsTupleType)
+            {
+                wasNotLiteral = false;
+                return AccessTupleFields(VisitExpression(tuple), temps, effects);
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        private BoundExpression RewriteTupleSingleOperator(TupleBinaryOperatorInfo.Single single,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType)
+        {
+            return _factory.Binary(single.Kind, boolType, left, right, single.MethodSymbolOpt);
```

Comment:
>_factory [](start = 19, length = 8)

I suspect that there may be some lowering not occurring here. For example, what happens if the operands are of type `decimal`?

Reply:
Lowering would turn the operator into a method invocation (e.g. when `single.MethodSymbolOpt` is non-null)

---
In reply to: [153607936](https://github.com/dotnet/roslyn/pull/23356#discussion_r153607936) [](ancestors = 153607936)

Reply:
Previously, I had a `VisitExpression` at the top-level (before returning from `VisitTupleBinaryOperator`). It turns out that is lowering some nodes twice. I've now fixed the lowering strategy to avoid double-lowering.

---
In reply to: [153608156](https://github.com/dotnet/roslyn/pull/23356#discussion_r153608156) [](ancestors = 153608156,153607936)